src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+{
+    /// <summary>
+    ///   Manages communication with other peers.
+    /// </summary>
+    public class SwarmService : ISwarmService
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(SwarmService));
+
+        /// <summary>
+        ///   The time to wait for a low level connection to be established.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 30 seconds.
+        /// </value>
+        public TimeSpan TransportConnectionTimeout = TimeSpan.FromSeconds(30);
+
+        /// <summary>
+        ///  The supported protocols.
+        /// </summary>
+        /// <remarks>
+        ///   Use sychronized access, e.g. <code>lock (protocols) { ... }</code>.
+        /// </remarks>
+        private List<IPeerProtocol> _protocols = new List<IPeerProtocol>
+        {
+            new Multistream1(),
+            new Secio1(),
+            new Identify1(),
+            new Mplex67()
+        };
+
+        /// <summary>
+        ///   Added to connection protocols when needed.
+        /// </summary>
+        private readonly Plaintext1 _plaintext1 = new Plaintext1();
+
+        private Peer _localPeer;
+
+        /// <summary>
+        ///   Raised when a listener is establihed.
+        /// </summary>
+        /// <remarks>
+        ///   Raised when <see cref="StartListeningAsync(MultiAddress)"/>
+        ///   succeeds.
+        /// </remarks>
+        public event EventHandler<Peer> ListenerEstablished;
+
+        /// <summary>
+        ///   Raised when a connection to another peer is established.
+        /// </summary>
+        public event EventHandler<PeerConnection> ConnectionEstablished;
+
+        /// <summary>
+        ///   Raised when a new peer is discovered for the first time.
+        /// </summary>
+        public event EventHandler<Peer> PeerDiscovered;
+
+        /// <summary>
+        ///   Raised when a peer's connection is closed.
+        /// </summary>
+        public event EventHandler<Peer> PeerDisconnected;
+
+        /// <summary>
+        ///   Raised when a peer should no longer be used.
+        /// </summary>
+        /// <remarks>
+        ///   This event indicates that the peer has been removed
+        ///   from the <see cref="KnownPeers"/> and should no longer
+        ///   be used.
+        /// </remarks>
+        public event EventHandler<Peer> PeerRemoved;
+
+        /// <summary>
+        ///   Raised when a peer cannot be connected to.
+        /// </summary>
+        public event EventHandler<Peer> PeerNotReachable;
+
+        /// <summary>
+        ///  The local peer.
+        /// </summary>
+        /// <value>
+        ///   The local peer must have an <see cref="Peer.Id"/> and
+        ///   <see cref="Peer.PublicKey"/>.
+        /// </value>
+        public Peer LocalPeer
+        {
+            get => _localPeer;
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use a constructor overloads that allows a more meaningful exception message to be provided.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)